File: PhotonShooter.cpp
/*
 *  PhotonShooter.cpp
 *  puls4r
 *
 *  Created by Camille on 06/03/08.
 *  Copyright 2008 __MyCompanyName__. All rights reserved.
 *
 */

#include "PhotonShooter.h"

const double PhotonShooter::EPSILON = 0.000001;
const double PhotonShooter::IRRADIANCE_AREA = 1.5;

void PhotonShooter::shoot(std::list<Light* > lights, std::list<Object * > objects){
	
	int lightNum = 0;
	
	for(std::list<Light* >::iterator iter = lights.begin(); iter != lights.end(); ++iter){
		lightNum++;
		
		int progress = 0;
		
		storedDirect = 0;
		storedIndirect = 0;
		storedCaustic = 0;
		
		while((storedDirect + storedIndirect + storedCaustic) < (maxPhotons / nbLights)){
			
			int done = (int)round(100 * ((storedDirect + storedIndirect + storedCaustic)+1) / (maxPhotons / nbLights));
			
			if(progress != done){
				progress = done;
				std::cout << "\r\tLight (" << lightNum << "/" << lights.size() << ") : " << progress << "%";
				std::cout.flush();
			}
			
			Vector randDir = randomDirection();
			Ray ray((*iter)->getSource(), &randDir);
			
			float energy[3] = {	(*iter)->getPower() * (*iter)->getColor()->getR(), 
								(*iter)->getPower() * (*iter)->getColor()->getG(), 
								(*iter)->getPower() * (*iter)->getColor()->getB()};
			
			currentRecursions = 0;
			shootPhoton(&ray, lights, objects, energy, true, false, false);
		}
		
		/* Photons energy scaling */
		if(storedDirect != 0){
			directEnlightment->scale_photon_power(1.0 / storedDirect);
		}
		
		if(storedIndirect != 0){
			indirectEnlightment->scale_photon_power(1.0 / storedIndirect);
		}
		
		if(storedCaustic != 0){
			caustics->scale_photon_power(1.0 / storedCaustic);
		}
		
		std::cout << std::endl;
	}
	
	std::cout << "---> Balancing direct photon map..." << std::endl;
	directEnlightment->balance();
	
	std::cout << "---> Balancing indirect photon map..." << std::endl;
	indirectEnlightment->balance();
	
	std::cout << "---> Balancing caustic photon map..." << std::endl;
	caustics->balance();
}

/*
 * Random emission from diffuse point light 
 * As described by Henrik Wann Jensen
 */
Vector PhotonShooter::randomDirection(void) const{
	
	double x, y, z;
	
	do{
		x = (((double)std::rand() / (double)RAND_MAX) * 2.0) - 1.0;
		y = (((double)std::rand() / (double)RAND_MAX) * 2.0) - 1.0;
		z = (((double)std::rand() / (double)RAND_MAX) * 2.0) - 1.0;
	}while( (x*x)+(y*y)+(z*z) > 1);
	
	Vector randDir(x, y, z);
	
	return randDir;
}

bool PhotonShooter::russianRoulette(double d) const{
	double e = ((double)std::rand() / (double)RAND_MAX); //e is in [0, 1]
	
	if(e < d){
		//Photon continues course
		return false;
	}
	else{
		//Absorption
		return true;
	}
}

void PhotonShooter::shootPhoton(Ray* ray, std::list<Light * > lights, std::list<Object * > objects, float energy[3], bool direct, bool indirect, bool caustic){
	
	//TODO split refraction coeff for all waves (r, g and b)
	currentRecursions++;
	
	Intersection* photonIntersection = getNearestIntersection(ray, objects);
	
	if(photonIntersection != 0){
		if(photonIntersection->getObject()->isReflecting()){
			
			if(russianRoulette(photonIntersection->getObject()->getKR()) || (currentRecursions >= MAX_RECURSIONS)){
				storePhoton(photonIntersection->getPoint(), ray->getDirection(), energy, direct, indirect, caustic);
			}
			else{
				Ray reflected = reflectedRay(ray, photonIntersection);
				shootPhoton(&reflected, lights, objects, energy, false, true, false);
			}
		}
		
		if(photonIntersection->getObject()->isRefracting()){
			
			if(russianRoulette(photonIntersection->getObject()->getKT()) || (currentRecursions >= MAX_RECURSIONS)){
				storePhoton(photonIntersection->getPoint(), ray->getDirection(), energy, direct, indirect, caustic);
			}
			else{
				Ray refracted = refractedRay(ray, photonIntersection, objects);
				shootPhoton(&refracted, lights, objects, energy, false, false, true);
			}
			
		}
		
		if(!photonIntersection->getObject()->isReflecting() && !photonIntersection->getObject()->isRefracting()){
			
			Color objectColor = photonIntersection->getObject()->getEnlightment()->getColor(photonIntersection->getPoint(), photonIntersection->getNormal(), ray, lights);
			scaleEnergy(energy, &objectColor);
			
			storePhoton(photonIntersection->getPoint(), ray->getDirection(), energy, direct, indirect, caustic);
			
			if(!russianRoulette(photonIntersection->getObject()->getKR()) && !(currentRecursions >= MAX_RECURSIONS)){
				Vector randDir = randomDirection();
				Ray reflected(photonIntersection->getPoint(), &randDir);
				shootPhoton(&reflected, lights, objects, energy, false, true, false);
			}
		}
	}
	
	delete(photonIntersection);
}

void PhotonShooter::storePhoton(Point* position, Vector* direction, float energy[3], bool direct, bool indirect, bool caustic){
	float* pos = position->toArray();
	float* dir = direction->toArray();
	
	if(direct){
		storedDirect++;
		directEnlightment->store(energy, pos, dir);
	}
	
	if(indirect){
		storedIndirect++;
		indirectEnlightment->store(energy, pos, dir);
	}
	
	if(caustic){
		storedCaustic++;
		caustics->store(energy, pos, dir);
	}
	
	free(pos);
	free(dir);
}

void PhotonShooter::scaleEnergy(float energy[3], Color* color){
	energy[0] *= (float)color->getR();
	energy[1] *= (float)color->getG();
	energy[2] *= (float)color->getB();
}

Color PhotonShooter::irradianceEstimate(Intersection* intersection){
	Color color;
	
	float *irradiance = (float*) malloc(3*sizeof(float));
	float* pos = intersection->getPoint()->toArray();
	float* normal = intersection->getNormal()->toArray();
	
	//Direct enlightment component
	irradiance[0] = 0;
	irradiance[1] = 0;
	irradiance[2] = 0;
	
	directEnlightment->irradiance_estimate(irradiance, pos, normal, IRRADIANCE_AREA, IRRADIANCE_PHOTON_NUMBER);
	
	color.setR(color.getR() + (double)irradiance[0]);
	color.setG(color.getG() + (double)irradiance[1]);
	color.setB(color.getB() + (double)irradiance[2]);			
	
	//Indirect enlightment component
	irradiance[0] = 0;
	irradiance[1] = 0;
	irradiance[2] = 0;
	
	indirectEnlightment->irradiance_estimate(irradiance, pos, normal, IRRADIANCE_AREA, IRRADIANCE_PHOTON_NUMBER);
	
	color.setR(color.getR() + (double)irradiance[0]);
	color.setG(color.getG() + (double)irradiance[1]);
	color.setB(color.getB() + (double)irradiance[2]);	
	
	//Caustics component
	irradiance[0] = 0;
	irradiance[1] = 0;
	irradiance[2] = 0;
	
	caustics->irradiance_estimate(irradiance, pos, normal, IRRADIANCE_AREA, IRRADIANCE_PHOTON_NUMBER);
	
	color.setR(color.getR() + (double)irradiance[0]);
	color.setG(color.getG() + (double)irradiance[1]);
	color.setB(color.getB() + (double)irradiance[2]);	
	
	
	free(irradiance);
	free(pos);
	free(normal);
	
	color.normalize();
	
	return color;
}

Intersection* PhotonShooter::getNearestIntersection(Ray* ray, std::list<Object * > objects, double _epsilon){
	
	Intersection* nearestIntersection = 0;
	
	for(std::list<Object* >::iterator iter = objects.begin(); iter != objects.end(); ++iter){
		
		Intersection* candidate = (*iter)->intersection(ray);
		
		if(candidate != 0){
			if((nearestIntersection == 0 || nearestIntersection->getT() > candidate->getT()) && (candidate->getT() > _epsilon)){
				if(nearestIntersection != 0){
					delete(nearestIntersection);
				}
				
				nearestIntersection = candidate;
			}
			else{
				delete(candidate);
			}
		}
		
	}
	
	return nearestIntersection;
}

Ray PhotonShooter::reflectedRay(Ray* ray, Intersection* intersection){
	Vector reflectDirection( (*ray->getDirection()) - ( ((*intersection->getNormal()) * 2.0) * ((*intersection->getNormal()) * ray->getDirection()) ) );
	reflectDirection.normalize();
	
	Point origin((reflectDirection * EPSILON) + intersection->getPoint());
	Ray reflected(&origin, &reflectDirection);
	
	return reflected;
}

Ray PhotonShooter::refractedRay(Ray* ray, Intersection* intersection, std::list<Object * > objects){
	
	Ray r1 = refractRay(ray, intersection, 1.0, intersection->getObject()->getN());
	Intersection* refractionIntersection = getNearestIntersection(&r1, objects);
	
	if(refractionIntersection != 0){
		if(refractionIntersection->getObject() == intersection->getObject()){
			refractionIntersection->getNormal()->invert();
			
			Ray r2 = refractRay(&r1, refractionIntersection, intersection->getObject()->getN(), 1.0);
			
			delete(refractionIntersection);
			return r2;
		}
		else{
			
			delete(refractionIntersection);
			return r1;
		}
	}
	else{
		return r1;
	}
}

Ray PhotonShooter::refractRay(Ray* ray, Intersection* intersection, double n1, double n2, double epsilon){
	
	double n = n2 / n1;
	double cosT = (*intersection->getNormal()) * ray->getDirection();
	double sinT2 = 1.0 - ((1.0  - cosT * cosT ) / ( n * n ));
	
	Vector refractDirection;
	
	if(sinT2 > 0){
		//refraction
		refractDirection = ( (( (*ray->getDirection()) - ( (*intersection->getNormal()) * cosT ) ) * (1.0 / n)) - ( (*intersection->getNormal()) * sqrt(sinT2) ));
	}
	else{
		//reflexion
		refractDirection = ( (*ray->getDirection()) - ( ((*intersection->getNormal()) * 2.0) * ((*intersection->getNormal()) * ray->getDirection()) ) );
	}
	
	Point origin((refractDirection * epsilon) + intersection->getPoint());
	Ray refracted(&origin, &refractDirection);
	
	return refracted;
}

File: trunk/Java/puls4r/enlightment/global/photonmapping/PhotonShooter.java
/**
 * 
 */
package puls4r.enlightment.global.photonmapping;

import javax.vecmath.Color3f;
import javax.vecmath.Point3d;
import javax.vecmath.Vector3d;

import puls4r.enlightment.Enlightment;
import puls4r.scene.Scene;
import puls4r.scene.objects.Light;
import puls4r.tracer.Intersection;
import puls4r.tracer.Ray;

/**
 * @author Camille
 *
 */
public class PhotonShooter {
	private static int MAX_RECURSIONS = 500;
	private static double IRRADIANCE_AREA = 1.5;
	private static int IRRADIANCE_PHOTON_NUMBER = 1000000;
	
	private int currentRecursions;
	
	private int storedDirect;
	private int storedIndirect;
	private int storedCaustic;
	
	private int maxPhotons;
	
	private PhotonMap directEnlightment;
	private PhotonMap indirectEnlightment;
	private PhotonMap caustics;
	
	private Scene scene;
	
	public PhotonShooter(Scene scene, int maxPhotons) {
		this.scene = scene;
		this.maxPhotons = maxPhotons;
		
		directEnlightment = new PhotonMap(maxPhotons);
		indirectEnlightment = new PhotonMap(maxPhotons);
		caustics = new PhotonMap(maxPhotons);
	}
	
	private Vector3d randomDirection(){
		double x, y, z;
		
		do{
			x = Math.random();
			y = Math.random();
			z = Math.random();
		}while( (x*x)+(y*y)+(z*z) > 1);
		
		Vector3d randDir = new Vector3d(x, y, z);
		
		return randDir;
	}
	
	private boolean russianRoulette(double d){
		double e = Math.random();
		
		if(e < d){
			//Photon continues course
			return false;
		}
		else{
			//Photon absorbed
			return true;
		}
	}
	
	private void storePhoton(Point3d position, Vector3d direction, float energy[], boolean direct, boolean indirect, boolean caustic){
		float pos[] = {(float) position.x, (float) position.y, (float) position.z};
		float dir[] = {(float) direction.x, (float) direction.y, (float) direction.z};
		
		if(direct){
			storedDirect++;
			directEnlightment.store(energy, pos, dir);
		}
		
		if(indirect){
			storedIndirect++;
			indirectEnlightment.store(energy, pos, dir);
		}
		
		if(caustic){
			storedCaustic++;
			caustics.store(energy, pos, dir);
		}
	}
	
	private void scaleEnergy(float energy[], Color3f color){
		energy[0] *= (float)color.x;
		energy[1] *= (float)color.y;
		energy[2] *= (float)color.z;
	}
	
	private void shootPhoton(Ray ray, float energy[], boolean direct, boolean indirect, boolean caustic){
		//TODO split refraction coeff for all waves (r, g and b)
		currentRecursions++;
		
		Intersection photonIntersection = scene.getNearestIntersection(ray);
		
		if(photonIntersection != null){
			if(photonIntersection.getObject().isReflecting()){
				
				if(russianRoulette(photonIntersection.getObject().getKR()) || (currentRecursions >= MAX_RECURSIONS)){
					storePhoton(photonIntersection.getPoint(), ray.getDirection(), energy, direct, indirect, caustic);
				}
				else{
					Ray reflected = scene.reflectedRay(ray, photonIntersection);
					shootPhoton(reflected, energy, false, true, false);
				}
			}
			
			if(photonIntersection.getObject().isRefracting()){
				
				if(russianRoulette(photonIntersection.getObject().getKT()) || (currentRecursions >= MAX_RECURSIONS)){
					storePhoton(photonIntersection.getPoint(), ray.getDirection(), energy, direct, indirect, caustic);
				}
				else{
					Ray refracted = scene.refractedRay(ray, photonIntersection);
					shootPhoton(refracted, energy, false, false, true);
				}
				
			}
			
			if(!photonIntersection.getObject().isReflecting() && !photonIntersection.getObject().isRefracting()){
				
				Color3f objectColor = photonIntersection.getObject().getEnlightment().getColor(photonIntersection.getPoint(), photonIntersection.getNormal(), ray, scene.getLights());
				scaleEnergy(energy, objectColor);
				
				storePhoton(photonIntersection.getPoint(), ray.getDirection(), energy, direct, indirect, caustic);
				
				if(!russianRoulette(photonIntersection.getObject().getKR()) && !(currentRecursions >= MAX_RECURSIONS)){
					Vector3d randDir = randomDirection();
					Ray reflected =  new Ray(photonIntersection.getPoint(), randDir);
					shootPhoton(reflected, energy, false, true, false);
				}
			}
		}
	}
	
	public void shoot(){
		int lightNum = 0;
		
		for(Light l : scene.getLights()){
			lightNum++;
			
			int progress = 0;
			
			storedDirect = 0;
			storedIndirect = 0;
			storedCaustic = 0;
			
			while((storedDirect + storedIndirect + storedCaustic) < (maxPhotons / scene.getLights().size())){
				
				int done = (int)Math.round(100 * ((storedDirect + storedIndirect + storedCaustic)+1) / (maxPhotons / scene.getLights().size()));
				
				if(progress != done){
					progress = done;
					System.out.println("\r\tLight (" + lightNum + "/" + scene.getLights().size() + ") : " + progress + "%");
				}
				
				Vector3d randDir = randomDirection();
				Ray ray = new Ray(l.getSource(), randDir);
				
				float energy[] = {	(float) (l.getPower() * l.getColor().x), 
									(float) (l.getPower() * l.getColor().y), 
									(float) (l.getPower() * l.getColor().z)};
				
				currentRecursions = 0;
				shootPhoton(ray, energy, true, false, false);
			}
			
			/* Photons energy scaling */
			if(storedDirect != 0){
				directEnlightment.scale_photon_power(1.0f / storedDirect);
			}
			
			if(storedIndirect != 0){
				indirectEnlightment.scale_photon_power(1.0f / storedIndirect);
			}
			
			if(storedCaustic != 0){
				caustics.scale_photon_power(1.0f / storedCaustic);
			}
			
			System.out.println("");
		}
		
		System.out.println("--> Balancing direct photon map...");
		directEnlightment.balance();
		
		System.out.println("--> Balancing indirect photon map...");
		indirectEnlightment.balance();
		
		System.out.println("--> Balancing caustic photon map...");
		caustics.balance();
	}
	
	public Color3f irradianceEstimate(Intersection intersection){
		Color3f color = new Color3f();
		
		float irradiance[] = new float[3];
		float pos[] = {(float) intersection.getPoint().x, (float) intersection.getPoint().y, (float) intersection.getPoint().z};
		float normal[] = {(float) intersection.getNormal().x, (float) intersection.getNormal().y, (float) intersection.getNormal().z};
		
		//Direct enlightment component
		irradiance[0] = 0;
		irradiance[1] = 0;
		irradiance[2] = 0;
		
		directEnlightment.irradiance_estimate(irradiance, pos, normal, (float) IRRADIANCE_AREA, IRRADIANCE_PHOTON_NUMBER);
		
		color.set(color.x + irradiance[0], color.y + irradiance[1], color.z + irradiance[2]);			
		
		//Indirect enlightment component
		irradiance[0] = 0;
		irradiance[1] = 0;
		irradiance[2] = 0;
		
		indirectEnlightment.irradiance_estimate(irradiance, pos, normal, (float) IRRADIANCE_AREA, IRRADIANCE_PHOTON_NUMBER);
		
		color.set(color.x + irradiance[0], color.y + irradiance[1], color.z + irradiance[2]);	
		
		//Caustics component
		irradiance[0] = 0;
		irradiance[1] = 0;
		irradiance[2] = 0;
		
		caustics.irradiance_estimate(irradiance, pos, normal, (float) IRRADIANCE_AREA, IRRADIANCE_PHOTON_NUMBER);
		
		color.set(color.x + irradiance[0], color.y + irradiance[1], color.z + irradiance[2]);
		Enlightment.normalize(color);
		
		return color;
	}
}
